ops/cluster.py

```diff
@@ -21,9 +22,579 @@
+    **kwargs : dict
+        Additional arguments passed to PHATE
+        
+    Returns:
+    --------
+    tuple
+        (DataFrame with PHATE coordinates, PHATE object)
+    """
+    # Initialize and run PHATE
+    p = phate.PHATE(
+        random_state=random_state,
+        n_jobs=n_jobs,
+        knn=knn,
+        knn_dist=metric,
+        **kwargs
+    )
+    
+    # Transform data
+    X_phate = p.fit_transform(df.values)
+    
+    # Create output DataFrame
+    df_phate = pd.DataFrame(
+        X_phate,
+        index=df.index,
+        columns=['PHATE_0', 'PHATE_1']
+    )
+    
+    return df_phate, p
+
+def run_leiden_clustering(weights, resolution=1.0, seed=42):
+    """
+    Run Leiden clustering on a weighted adjacency matrix.
+    
+    Parameters:
+    -----------
+    weights : numpy.ndarray
+        Weighted adjacency matrix
+    resolution : float, default=1.0
+        Resolution parameter for Leiden clustering
+    seed : int, default=42
+        Random seed for reproducibility
+    
+    Returns:
+    --------
+    list
+        Cluster assignments
+    """
+    # Force symmetry by averaging with transpose
+    weights_symmetric = (weights + weights.T) / 2
+    
+    # Create graph from symmetrized weights
+    g = Graph().Weighted_Adjacency(
+        matrix=weights_symmetric.tolist(),
+        mode='undirected'
+    )
+    
+    # Run Leiden clustering
+    partition = leidenalg.find_partition(
+        g,
+        partition_type=leidenalg.RBConfigurationVertexPartition,
+        weights=g.es['weight'],
+        n_iterations=-1,
+        seed=seed,
+        resolution_parameter=resolution
+    )
+    
+    return partition.membership
+
+def phate_leiden_pipeline(df, resolution=1.0, phate_kwargs=None):
+    """
+    Run complete PHATE and Leiden clustering pipeline.
+    
+    Parameters:
+    -----------
+    df : pandas.DataFrame
+        Input data matrix
+    resolution : float, default=1.0
+        Resolution parameter for Leiden clustering
+    phate_kwargs : dict, optional
+        Additional arguments for PHATE
+        
+    Returns:
+    --------
+    pandas.DataFrame
+        DataFrame with PHATE coordinates and cluster assignments
+    """
+    # Default PHATE parameters
+    if phate_kwargs is None:
+        phate_kwargs = {}
+    
+    # Run PHATE
+    df_phate, p = run_phate(df, **phate_kwargs)
+    
+    # Get weights from PHATE
+    weights = np.asarray(p.graph.diff_op.todense())
+    
+    # Run Leiden clustering
+    clusters = run_leiden_clustering(weights, resolution=resolution)
+    
+    # Add clusters to results
+    df_phate['cluster'] = clusters
+
+    # Sort by cluster
+    df_phate = df_phate.sort_values('cluster')
+
+    # Print number of clusters and average cluster size
+    print(f"Number of clusters: {df_phate['cluster'].nunique()}")
+    print(f"Average cluster size: {df_phate['cluster'].value_counts().mean():.2f}")
+    
+    return df_phate
+
+# REMOVE ALL BELOW?
```

Comment:
Maybe you are able to answer this now that OPS is pretty much complete? Or we can get back to this later when migrating code to Brieflow.

Reply:
This isn't used so we can remove it, yes.